---
title: "Symbiodiniaceae LULU"
author: "Vera Emelianenko"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library("tidyverse")
library("lulu")
## Empty workspace ##
rm(list=ls())
packageVersion("lulu")
```

# Filter by length 
Filter out the ASVs that have less than 250 nucleotides length
```
> cat input/ASVs_run1-2-3.fasta | seqkit seq --min-len 250 > ASVs_run1-2-3_minlen250.fasta
> seqkit seq -n ASVs_run1-2-3_minlen250.fasta > ASVs_minlen250_ids.txt
```

```{r}
ASVs_minlen_250_id <- read.csv('./ASVs_minlen250_ids.txt', sep='\t', header=FALSE, as.is=TRUE)

ASVs_counts <-  data.table::fread("./input/ASVs_counts_run1-2-3.txt") # 1541 rows
any(is.na(ASVs_counts))

# Sanity check to get occurence of each ASV
row.names(ASVs_counts) <- ASVs_counts$V1
ASVs_counts_noname <- ASVs_counts %>% select(-V1)
ASVs_counts_summarized <- ASVs_counts_noname  %>%
      rowwise() %>%
      mutate(RowSum = sum(c_across(everything()))) %>% dplyr::select(RowSum)
ASVs_counts_summarized$ASV <- rownames(ASVs_counts_noname)

ASV_counts_minlen250 <- left_join(ASVs_minlen_250_id, ASVs_counts, by=join_by(V1 == V1)) %>% 
  dplyr::rename(ASV=V1) #1434 rows

# any(is.na(ASV_counts_minlen250))  # FALSE


write.table(ASV_counts_minlen250, './ASVs_counts_run1-2-3_minlen250.txt', sep="\t", quote=F, row.names = F)

# Sanity check again
row.names(ASV_counts_minlen250) <- ASV_counts_minlen250$ASV
ASVs_counts_minlen250_noname <- ASV_counts_minlen250 %>% select(-ASV)
ASVs_counts_minlen250_summarized <- ASVs_counts_minlen250_noname %>%
      rowwise() %>%
      mutate(RowSum = sum(c_across(everything()))) %>% dplyr::select(RowSum)
ASVs_counts_minlen250_summarized$ASV <- rownames(ASV_counts_minlen250)

# check that sums are the same 
left_join(ASVs_counts_minlen250_summarized, ASVs_counts_summarized,by=join_by(ASV == ASV)) %>% dplyr::filter(RowSum.x!=RowSum.y)
```

# Merge counts between the runs and the aliquots of sediment and algae

```{r}
# replace file extention in column names
colnames(ASV_counts_minlen250) <- gsub("_R1_001.fastq.gz", "", colnames(ASV_counts_minlen250))

colnames(ASV_counts_minlen250) 

write.table(as.data.frame(unlist(colnames(ASV_counts_minlen250))), './colnames.txt', sep="\t", quote=F, row.names = F, col.names = F) 
```

Summarise the columns "manually":

```{r}
ASV_counts_mutated <- ASV_counts_minlen250 %>% 
  dplyr::mutate(
    `Algae1-1` = `A1_1a1_ITS2_S179` + `A1_1a2_ITS2_S180`+ `A1-1a1-ITS2_S179_L001` + `A1-1a2-ITS2_S180_L001`, 
    `Algae1-2` = `A1_2a1_ITS2_S181` + `A1_2a2_ITS2_S182` + `A1-2a1-ITS2_S181_L001` + `A1-2a2-ITS2_S182_L001`, 
    `Algae1-5` = `A1_5a1_ITS2_S183` + `A1_5a2_ITS2_S184` + `A1-5a1-ITS2_S183_L001` + `A1-5a2-ITS2_S184_L001`,
    `Algae1-6` = `A1_6a1_ITS2_S185` + `A1_6a2_ITS2_S186` + `A1-6a1-ITS2_S185_L001` + `A1-6a2-ITS2_S186_L001`,
    `Algae2-2` = `A2_2a5_ITS2_S191` + `A2_2a6_ITS2_S192` + `A2-2a5-ITS2_S191_L001` + `A2-2a6-ITS2_S192_L001`,
    `Algae2-5` = `A2_5a3_ITS2_S187` + `A2_5a4_ITS2_S188` + `A2-5a3-ITS2_S187_L001` + `A2-5a4-ITS2_S188_L001`,
    `Algae2-6` = `A2_6a3_ITS2_S189` + `A2_6a4_ITS2_S190` + `A2-6a3-ITS2_S189_L001` + `A2-6a4-ITS2_S190_L001`,
    `Algae2-7` = `A2_7a3_ITS2_S193` + `A2_7a4_ITS2_S194` + `A2-7a3-ITS2_S193_L001` + `A2-7a4-ITS2_S194_L001`)%>%   dplyr::mutate(
      `Stichodactyla1-1` = `An1_ITS2_S125` + `An1-ITS2_S125_L001`, 
      `Stichodactyla1-2` = `An2_ITS2_S126` +  `An2-ITS2_S126_L001`) %>% 
  dplyr::mutate (
    `Porites1-12` = `C1_12_corr_ITS2_S112` + `C1-12-corr-ITS2_S112_L001` + `C1-1_S71_L001`,
    `Goniastrea1-14` = `C1_14_ITS2_S111` + `C1-14-ITS2_S111_L001`,
    `Goniastrea1-1` = `C1_1_ITS2_S107` + `C1-1-ITS2_S107_L001`,
    `Montipora1-2` = `C1_2_corr_ITS2_S114` + `C1-2-corr-ITS2_S114_L001`,
    `Goniastrea1-3` = `C1_3_ITS2_S108` + `C1-3-ITS2_S108_L001`,
    `Goniastrea1-5` = `C1_5_ITS2_S109` + `C1-5-ITS2_S109_L001`,
    `Porites1-6` = `C1_6_ITS2_S113` + `C1-6-ITS2_S113_L001`,
    `Goniastrea1-9` = `C1_9_ITS2_S110` + `C1-9-ITS2_S110_L001`,
    `Goniastrea2-14` = `C2_14b_ITS2_S121` + `C2-14b-ITS2_S121_L001`,
    `Goniastrea2-16` = `C2_16_ITS2_S122` + `C2-16-ITS2_S122_L001`,
    `Goniastrea2-18` = `C2_18_ITS2_S123` + `C2-18-ITS2_S123_L001`,
    `Goniastrea2-19` = `C2_19_ITS2_S124` + `C2-19-ITS2_S124_L001`,
    `Porites2-1` = `C2_1_ITS2_S115` + `C2-1-ITS2_S115_L001`,
    `Goniastrea2-2` = `C2_2_ITS2_S120` + `C2-2-ITS2_S120_L001`,
    `Porites2-3` = `C2_3_ITS2_S116` + `C2-3-ITS2_S116_L001`,
    `Montipora2-6` = `C2_6_ITS2_S117` + `C2-6-ITS2_S117_L001`,
    `Porites2-7` = `C2_7_ITS2_S118` + `C2-7-ITS2_S118_L001`,
    `Porites2-9` = `C2_9_ITS2_S119` + `C2-9-ITS2_S119_L001`) %>% 
  dplyr::mutate(
    `Amphisorus1-10` = `F1_10_ITS2_S141` + `F1-10-ITS2_S141_L001`,
    `Sorites1-12` = `F1_12_ITS2_S149` + `F1-12-ITS2_S149_L001`,
    `Sorites1-13` = `F1_13_ITS2_S150` + `F1-13-ITS2_S150_L001`,
    `Sorites1-14` = `F1_14_ITS2_S151` + `F1-14-ITS2_S151_L001`,
    `Amphisorus1-2` = `F1_2_ITS2_S136` + `F1-2-ITS2_S136_L001`,
    `Amphisorus1-3` = `F1_3_ITS2_S137` + `F1-3-ITS2_S137_L001`,
    `Amphisorus1-4` = `F1_4_ITS2_S138` + `F1-4-ITS2_S138_L001`,
    `Amphisorus1-5` = `F1_5_ITS2_S139` + `F1-5-ITS2_S139_L001`,
    `Amphisorus1-7` = `F1_7_ITS2_S140` + `F1-7-ITS2_S140_L001`,
    `Sorites1-9` = `F1_9_ITS2_S148` + `F1-9-ITS2_S148_L001`,
    `Amphisorus2-1` = `F2_1_ITS2_S142` + `F2-1-ITS2_S142_L001`,
    `Sorites2-21` = `F2_21_ITS2_S152` + `F2-21-ITS2_S152_L001`,
    `Sorites2-22` = `F2_22_ITS2_S153` + `F2-22-ITS2_S153_L001`,
    `Sorites2-29` = `F2_29_ITS2_S154` + `F2-29-ITS2_S154_L001`,
    `Amphisorus2-2` = `F2_2_ITS2_S143` + `F2-2-ITS2_S143_L001`,
    `Amphisorus2-30` = `F2_30_ITS2_S146` + `F2-30-ITS2_S146_L001`,
    `Amphisorus2-31` = `F2_31_ITS2_S147` + `F2-31-ITS2_S147_L001`,
    `Amphisorus2-3` = `F2_3_ITS2_S144` + `F2-3-ITS2_S144_L001`,
    `Amphisorus2-4` = `F2_4_ITS2_S145` + `F2-4-ITS2_S145_L001`, 
    `Amphisorus2-5` = `F2-5_S72_L001`, 
    `Amphisorus2-6` = `F2-6_S73_L001`) %>% 
  dplyr::mutate(
    `Sediment1-5` = `S1_5a1_ITS2_S155` + `S1_5a2_ITS2_S156` + `S1-5a1-ITS2_S155_L001` + `S1-5a2-ITS2_S156_L001`,
    `Sediment1-6` = `S1_6a1_ITS2_S157` + `S1_6a2_ITS2_S158` + `S1-6a1-ITS2_S157_L001` + `S1-6a2-ITS2_S158_L001`,
    `Sediment1-7` = `S1_7a1_ITS2_S159` + `S1_7a2_ITS2_S160` + `S1-7a1-ITS2_S159_L001` + `S1-7a2-ITS2_S160_L001`,
    `Sediment1-8` = `S1_8a3_ITS2_S161` + `S1_8a4_ITS2_S162` + `S1-8a3-ITS2_S161_L001` + `S1-8a4-ITS2_S162_L001`,
    `Sediment2-4` = `S2_4a3_ITS2_S163` + `S2_4a4_ITS2_S164` + `S2-4a3-ITS2_S163_L001` + `S2-4a4-ITS2_S164_L001`,
    `Sediment2-5` = `S2_5a1_ITS2_S165` + `S2_5a2_ITS2_S166` + `S2-5a1-ITS2_S165_L001` + `S2-5a2-ITS2_S166_L001`,
    `Sediment2-6` = `S2_6a1_ITS2_S167` + `S2_6a2_ITS2_S168` + `S2-6a1-ITS2_S167_L001` + `S2-6a2-ITS2_S168_L001`,
    `Sediment2-9` = `S2_9a1_ITS2_S169` + `S2_9a2_ITS2_S170` + `S2-9a1-ITS2_S169_L001` + `S2-9a2-ITS2_S170_L001`) %>% 
  dplyr::mutate(
    `Water1-1` = `W1_1_ITS2_S171` + `W1-1-ITS2_S171_L001`,
    `Water1-2` = `W1_2_ITS2_S172` + `W1-2-ITS2_S172_L001`,
    `Water1-3` = `W1_3_ITS2_S173` + `W1-3-ITS2_S173_L001`,
    `Water1-4` = `W1_4_ITS2_S174` + `W1-4-ITS2_S174_L001`,
    `Water2-5` = `W2_5_ITS2_S175` + `W2-5-ITS2_S175_L001`,
    `Water2-6` = `W2_6_ITS2_S176` + `W2-6-ITS2_S176_L001`,
    `Water2-7` = `W2_7_ITS2_S177` + `W2-7-ITS2_S177_L001`,
    `Water2-8` = `W2_8_ITS2_S178` + `W2-8-ITS2_S178_L001`) %>% 
  dplyr::mutate(
    `Tridacna1` = `TN1a_ITS2_S127` + `TN1a-ITS2_S127_L001`,
  `Tridacna2` = `TN2a_ITS2_S128` + `TN2a-ITS2_S128_L001`,
  `Tridacna3` = `TN3a_ITS2_S129` + `TN3a-ITS2_S129_L001`,
  `Tridacna4` = `TN4a_ITS2_S130` + `TN4a-ITS2_S130_L001`,
  `Tridacna5` = `TN5a_ITS2_S131` + `TN5a-ITS2_S131_L001`,
  `Tridacna6` = `TN6a_ITS2_S132` + `TN6a-ITS2_S132_L001`,
  `Tridacna7` = `TN7a_ITS2_S133` + `TN7a-ITS2_S133_L001`,
  `Tridacna8` = `TN8a_ITS2_S134` + `TN8a-ITS2_S134_L001`,
  `Tridacna9` = `TN9a_ITS2_S135` + `TN9a-ITS2_S135_L001`)

# Sanity check
any(is.na(ASV_counts_mutated))
```

Select only renamed columns and check
```{r}
ASV_counts_renamed <- ASV_counts_mutated %>% dplyr::select(ASV, `Algae1-1`:`Tridacna9`)

any(is.na(ASV_counts_renamed ))

# Sanity check 
row.names(ASV_counts_renamed) <- ASV_counts_renamed$ASV
ASV_counts_renamed_noname <- ASV_counts_renamed %>% select(-ASV)
ASV_counts_renamed_summarized <- ASV_counts_renamed_noname %>%
      rowwise() %>%
      mutate(RowSum = sum(c_across(everything()))) %>% dplyr::select(RowSum)
ASV_counts_renamed_summarized$ASV <- rownames(ASV_counts_minlen250)

# check that sums are the same 
left_join(ASVs_counts_minlen250_summarized, ASV_counts_renamed_summarized,by=join_by(ASV == ASV)) %>% dplyr::filter(RowSum.x!=RowSum.y)

# Write the table down
write.table(ASV_counts_renamed, './output/ASV_counts_renamed.txt', sep="\t", quote=F, row.names = F, col.names = T) 

```

# Run lulu 
## Prepare the data and libraries 
```{r}
# uncoment if need to install
# library("devtools")
#install_github("tobiasgf/lulu")
library("lulu")
```

The code below is taken from https://github.com/tobiasgf/lulu with file name modifications. 
We used LULU R package to reduce the number of ASVs, following the workflow suggested in ([LULU GitHub](https://github.com/tobiasgf/lulu)).

It removed taxonomically relative entities such as *Halluxium* as errors of other variants at the default parameters (minimum_match = 84, minimum_relative_cooccurence = 0.95 or 0.90) so they were set to be more strict (minimum_match = 95, minimum_relative_cooccurence = 0.95). It makes sense as 84% similarity is generally not a very high percentage and might remove the ASVs that coocure due to ecological reasons (for example, if some organism is only present in the water column, and another organism is also present only in the water column, they will both be found in water samples and thus coocure even if they are different biological entities).

The file with representative sequences is `ASVs_run1-2-3_minlen250.fasta` (headers are ASV_1, ASV_2... ASV_1458), and the OTU table is `ASVs_counts_run1-2-3_minlen250.txt`.

Produce a match list with blastn
```
# First produce a blastdatabase with the OTUs
makeblastdb -in ASVs_run1-2-3_minlen250.fasta -parse_seqids -dbtype nucl


# Then blast the OTUs against the database
# blastn: 2.12.0+
blastn -db ASVs_run1-2-3_minlen250.fasta -outfmt '6 qseqid sseqid pident' -out match_list_minlen250.txt -qcov_hsp_perc 80 -perc_identity 95 -query ASVs_run1-2-3_minlen250.fasta


# Alternatively, can use vsearch - NOT used for final data analysis
vsearch --usearch_global ASVs_run1-2-3_minlen250.fasta --db ASVs_run1-2-3_minlen250.fasta --self --id .90 --iddef 1 --userout match_list.txt -userfields query+target+id --maxaccepts 0 --query_cov .9 --maxhits 10
#  blastn/vsearch give slightly different results and can change clustering 
```
`match_list_minlen250.txt`/ `match_list.txt` format:
```
ASV_1   ASV_1   100.000
ASV_1   ASV_461 100.000
```


```{r}
# read the otu counts table 

otutab <- read.csv('./ASV_counts_renamed.txt', sep='\t', header=TRUE, as.is=TRUE, row.names = 1)

# read matchlist
matchlist <- read.table("match_list_minlen250.txt", header=FALSE, as.is=TRUE, stringsAsFactors=FALSE)
```

## Run lulu curation 
Run LULU with 95% sequence similarity and 95% co-occurrence of ASVs across samples.
```{r results='hide'}
curated_result <- lulu(otutab, matchlist, minimum_ratio_type = "min", minimum_ratio = 1, minimum_match = 95, minimum_relative_cooccurence = 0.95)
```

Explore the data
```{r}
# Number of OTUs discarded
curated_result$discarded_count #512

# Number of OTUs retained 
curated_result$curated_count #922
```
```{r}
# new modified table 
head(curated_result$curated_table) 
curated_table <- as.data.frame(curated_result$curated_table)
```


Write the ids of retained OTUs
```{r}
write.table(curated_result$curated_table, './output/ASVs_counts_curated.txt', sep="\t", quote=F)
write.table(as.data.frame(curated_result$curated_otus), 'curated_otus_ids.txt', sep="\t", quote=F, col.names = F, row.names = F)
```

Make a fasta file with retained OTUs with seqkit

```
seqkit grep -f curated_otus_ids.txt ASVs_run1-2-3_minlen250.fasta -o ./output/ASVs_curated.fasta
```

Output files after running LULU:

`ASVs_counts_curated.txt` - new curated ASV counts table

`ASVs_curated.fasta` - new curated ASV sequences (subset of original ASVs.fasta)

Unmerged counts table to visualize different aliquotes 

```{r}
ASV_counts_minlen250 <- data.table::fread('./ASVs_counts_run1-2-3_minlen250.txt')
  
# replace file extention in column names
colnames(ASV_counts_minlen250) <- gsub("_R1_001.fastq.gz", "", colnames(ASV_counts_minlen250))

ASV_counts_mutated_unmerged <- ASV_counts_minlen250 %>% 
  dplyr::mutate(
    `Algae1-1a1` = `A1_1a1_ITS2_S179` + `A1-1a1-ITS2_S179_L001`, 
    `Algae1-1a2` = `A1_1a2_ITS2_S180` + `A1-1a2-ITS2_S180_L001`, 
    
    `Algae1-2a1` = `A1_2a1_ITS2_S181` + `A1-2a1-ITS2_S181_L001` , 
    `Algae1-2a2` = `A1_2a2_ITS2_S182` + `A1-2a2-ITS2_S182_L001`, 
    
    `Algae1-5a1` = `A1_5a1_ITS2_S183` + `A1-5a1-ITS2_S183_L001`,
    `Algae1-5a2` = `A1_5a2_ITS2_S184` + `A1-5a2-ITS2_S184_L001`,
    
    `Algae1-6a1` = `A1_6a1_ITS2_S185` + `A1-6a1-ITS2_S185_L001` ,
    `Algae1-6a2` = `A1_6a2_ITS2_S186` + `A1-6a2-ITS2_S186_L001`,
    
    `Algae2-2a5` = `A2_2a5_ITS2_S191` + `A2-2a5-ITS2_S191_L001`,
    `Algae2-2a6` = `A2_2a6_ITS2_S192` + `A2-2a6-ITS2_S192_L001`,
    
    `Algae2-5a3` = `A2_5a3_ITS2_S187` + `A2-5a3-ITS2_S187_L001`,
    `Algae2-5a4` = `A2_5a4_ITS2_S188` + `A2-5a4-ITS2_S188_L001`,
    
    `Algae2-6a3` = `A2_6a3_ITS2_S189` + `A2-6a3-ITS2_S189_L001`,
    `Algae2-6a4` = `A2_6a4_ITS2_S190` + `A2-6a4-ITS2_S190_L001`,
    
    `Algae2-7a3` = `A2_7a3_ITS2_S193` + `A2-7a3-ITS2_S193_L001`,
    `Algae2-7a4` = `A2_7a4_ITS2_S194` + `A2-7a4-ITS2_S194_L001`)%>%   
  dplyr::mutate(
      `Stichodactyla1-1` = `An1_ITS2_S125` + `An1-ITS2_S125_L001`, 
      `Stichodactyla1-2` = `An2_ITS2_S126` +  `An2-ITS2_S126_L001`) %>% 
  dplyr::mutate (
    `Porites1-12` = `C1_12_corr_ITS2_S112` + `C1-12-corr-ITS2_S112_L001` + `C1-1_S71_L001`,
    `Goniastrea1-14` = `C1_14_ITS2_S111` + `C1-14-ITS2_S111_L001`,
    `Goniastrea1-1` = `C1_1_ITS2_S107` + `C1-1-ITS2_S107_L001`,
    `Montipora1-2` = `C1_2_corr_ITS2_S114` + `C1-2-corr-ITS2_S114_L001`,
    `Goniastrea1-3` = `C1_3_ITS2_S108` + `C1-3-ITS2_S108_L001`,
    `Goniastrea1-5` = `C1_5_ITS2_S109` + `C1-5-ITS2_S109_L001`,
    `Porites1-6` = `C1_6_ITS2_S113` + `C1-6-ITS2_S113_L001`,
    `Goniastrea1-9` = `C1_9_ITS2_S110` + `C1-9-ITS2_S110_L001`,
    `Goniastrea2-14` = `C2_14b_ITS2_S121` + `C2-14b-ITS2_S121_L001`,
    `Goniastrea2-16` = `C2_16_ITS2_S122` + `C2-16-ITS2_S122_L001`,
    `Goniastrea2-18` = `C2_18_ITS2_S123` + `C2-18-ITS2_S123_L001`,
    `Goniastrea2-19` = `C2_19_ITS2_S124` + `C2-19-ITS2_S124_L001`,
    `Porites2-1` = `C2_1_ITS2_S115` + `C2-1-ITS2_S115_L001`,
    `Goniastrea2-2` = `C2_2_ITS2_S120` + `C2-2-ITS2_S120_L001`,
    `Porites2-3` = `C2_3_ITS2_S116` + `C2-3-ITS2_S116_L001`,
    `Montipora2-6` = `C2_6_ITS2_S117` + `C2-6-ITS2_S117_L001`,
    `Porites2-7` = `C2_7_ITS2_S118` + `C2-7-ITS2_S118_L001`,
    `Porites2-9` = `C2_9_ITS2_S119` + `C2-9-ITS2_S119_L001`) %>% 
  dplyr::mutate(
    `Amphisorus1-10` = `F1_10_ITS2_S141` + `F1-10-ITS2_S141_L001`,
    `Sorites1-12` = `F1_12_ITS2_S149` + `F1-12-ITS2_S149_L001`,
    `Sorites1-13` = `F1_13_ITS2_S150` + `F1-13-ITS2_S150_L001`,
    `Sorites1-14` = `F1_14_ITS2_S151` + `F1-14-ITS2_S151_L001`,
    `Amphisorus1-2` = `F1_2_ITS2_S136` + `F1-2-ITS2_S136_L001`,
    `Amphisorus1-3` = `F1_3_ITS2_S137` + `F1-3-ITS2_S137_L001`,
    `Amphisorus1-4` = `F1_4_ITS2_S138` + `F1-4-ITS2_S138_L001`,
    `Amphisorus1-5` = `F1_5_ITS2_S139` + `F1-5-ITS2_S139_L001`,
    `Amphisorus1-7` = `F1_7_ITS2_S140` + `F1-7-ITS2_S140_L001`,
    `Sorites1-9` = `F1_9_ITS2_S148` + `F1-9-ITS2_S148_L001`,
    `Amphisorus2-1` = `F2_1_ITS2_S142` + `F2-1-ITS2_S142_L001`,
    `Sorites2-21` = `F2_21_ITS2_S152` + `F2-21-ITS2_S152_L001`,
    `Sorites2-22` = `F2_22_ITS2_S153` + `F2-22-ITS2_S153_L001`,
    `Sorites2-29` = `F2_29_ITS2_S154` + `F2-29-ITS2_S154_L001`,
    `Amphisorus2-2` = `F2_2_ITS2_S143` + `F2-2-ITS2_S143_L001`,
    `Amphisorus2-30` = `F2_30_ITS2_S146` + `F2-30-ITS2_S146_L001`,
    `Amphisorus2-31` = `F2_31_ITS2_S147` + `F2-31-ITS2_S147_L001`,
    `Amphisorus2-3` = `F2_3_ITS2_S144` + `F2-3-ITS2_S144_L001`,
    `Amphisorus2-4` = `F2_4_ITS2_S145` + `F2-4-ITS2_S145_L001`, 
    `Amphisorus2-5` = `F2-5_S72_L001`, 
    `Amphisorus2-6` = `F2-6_S73_L001`) %>% 
  dplyr::mutate(
    # Sediment1
    `Sediment1-5a1` = `S1_5a1_ITS2_S155` + `S1-5a1-ITS2_S155_L001`,
    `Sediment1-5a2` = `S1_5a2_ITS2_S156` + `S1-5a2-ITS2_S156_L001`,
    `Sediment1-6a1` = `S1_6a1_ITS2_S157` + `S1-6a1-ITS2_S157_L001`,
    `Sediment1-6a2` = `S1_6a2_ITS2_S158` + `S1-6a2-ITS2_S158_L001`,
    `Sediment1-7a1` = `S1_7a1_ITS2_S159` + `S1-7a1-ITS2_S159_L001`,
    `Sediment1-7a2` = `S1_7a2_ITS2_S160` + `S1-7a2-ITS2_S160_L001`,
    `Sediment1-8a3` = `S1_8a3_ITS2_S161` + `S1-8a3-ITS2_S161_L001`,
    `Sediment1-8a4` = `S1_8a4_ITS2_S162` + `S1-8a4-ITS2_S162_L001`,
    
    # Sediment2
    `Sediment2-4a3` = `S2_4a3_ITS2_S163` + `S2-4a3-ITS2_S163_L001`,
    `Sediment2-4a4` = `S2_4a4_ITS2_S164` + `S2-4a4-ITS2_S164_L001`,
    `Sediment2-5a1` = `S2_5a1_ITS2_S165` + `S2-5a1-ITS2_S165_L001`,
    `Sediment2-5a2` = `S2_5a2_ITS2_S166` + `S2-5a2-ITS2_S166_L001`,
    `Sediment2-6a1` = `S2_6a1_ITS2_S167` + `S2-6a1-ITS2_S167_L001`,
    `Sediment2-6a2` = `S2_6a2_ITS2_S168` + `S2-6a2-ITS2_S168_L001`,
    `Sediment2-9a1` = `S2_9a1_ITS2_S169` + `S2-9a1-ITS2_S169_L001`,
    `Sediment2-9a2` = `S2_9a2_ITS2_S170` + `S2-9a2-ITS2_S170_L001`
)  %>% 
  dplyr::mutate(
    `Water1-1` = `W1_1_ITS2_S171` + `W1-1-ITS2_S171_L001`,
    `Water1-2` = `W1_2_ITS2_S172` + `W1-2-ITS2_S172_L001`,
    `Water1-3` = `W1_3_ITS2_S173` + `W1-3-ITS2_S173_L001`,
    `Water1-4` = `W1_4_ITS2_S174` + `W1-4-ITS2_S174_L001`,
    `Water2-5` = `W2_5_ITS2_S175` + `W2-5-ITS2_S175_L001`,
    `Water2-6` = `W2_6_ITS2_S176` + `W2-6-ITS2_S176_L001`,
    `Water2-7` = `W2_7_ITS2_S177` + `W2-7-ITS2_S177_L001`,
    `Water2-8` = `W2_8_ITS2_S178` + `W2-8-ITS2_S178_L001`) %>% 
  dplyr::mutate(
    `Tridacna1` = `TN1a_ITS2_S127` + `TN1a-ITS2_S127_L001`,
  `Tridacna2` = `TN2a_ITS2_S128` + `TN2a-ITS2_S128_L001`,
  `Tridacna3` = `TN3a_ITS2_S129` + `TN3a-ITS2_S129_L001`,
  `Tridacna4` = `TN4a_ITS2_S130` + `TN4a-ITS2_S130_L001`,
  `Tridacna5` = `TN5a_ITS2_S131` + `TN5a-ITS2_S131_L001`,
  `Tridacna6` = `TN6a_ITS2_S132` + `TN6a-ITS2_S132_L001`,
  `Tridacna7` = `TN7a_ITS2_S133` + `TN7a-ITS2_S133_L001`,
  `Tridacna8` = `TN8a_ITS2_S134` + `TN8a-ITS2_S134_L001`,
  `Tridacna9` = `TN9a_ITS2_S135` + `TN9a-ITS2_S135_L001`)


any(is.na(ASV_counts_mutated_unmerged))

ASV_counts_renamed_unmerged <- ASV_counts_mutated_unmerged %>% dplyr::select(ASV, `Algae1-1a1`:`Tridacna9`)

# Sanity check again
row.names(ASV_counts_minlen250) <- ASV_counts_minlen250$ASV
ASVs_counts_minlen250_noname <- ASV_counts_minlen250 %>% select(-ASV)
ASVs_counts_minlen250_summarized <- ASVs_counts_minlen250_noname %>%
      rowwise() %>%
      mutate(RowSum = sum(c_across(everything()))) %>% dplyr::select(RowSum)
ASVs_counts_minlen250_summarized$ASV <- rownames(ASV_counts_minlen250)

#Sanity check 
row.names(ASV_counts_renamed_unmerged) <- ASV_counts_renamed_unmerged$ASV
ASV_counts_renamed_unmerged_noname <- ASV_counts_renamed_unmerged %>% select(-ASV)
ASV_counts_renamed_unmerged_summarized <- ASV_counts_renamed_unmerged_noname %>%
      rowwise() %>%
      mutate(RowSum = sum(c_across(everything()))) %>% dplyr::select(RowSum)
ASV_counts_renamed_unmerged_summarized$ASV <- rownames(ASV_counts_minlen250)

# check that sums are the same 
left_join(ASVs_counts_minlen250_summarized, ASV_counts_renamed_unmerged_summarized,by=join_by(ASV == ASV)) %>% dplyr::filter(RowSum.x!=RowSum.y)

write.table(ASV_counts_renamed_unmerged, './output/ASV_counts_renamed_unmerged.txt', sep="\t", quote=F, row.names = F, col.names = T) 
```

```

